source/includes/_errors.md

```diff
@@ -1,5 +1,18 @@
 # Errors
 
+The Smart-EDi API uses the following error codes:
+
+Error Code | Meaning
+---------- | -------
+400 | Bad Request -- Your request is invalid.
+404 | Not Found -- The specified record could not be found. ??? just a load of html, need api error handling ???
+500 | Internal Server Error -- We had a problem with our server. Try again later.
+
```

Comment:
I don't think we need to specify error codes and their meaning. Just, SmartEdi uses the standard API error codes or something. We will have more than just 3.
The API will respond with the right message when something happens.

Reply:
Where should i make a ticket to implement the error codes? M&E trello board?

Reply:
Yes, Trello.

Start with this:

```
module API
  class BaseController < ActionController::Base
    include Pagy::Backend
    
    ERRORS = {
      server_error: "Something happened on our end, it's us. Please try again and if the issue persists, get in touch.",
      missing_token: "Couldn't find your authentication token in the header request.",
      invalid_or_expired_token: "Couldn't find an authority for this token. Your token is either invalid or expired.",
      per_page_exceeded: "'per_page' param value exceeds the maxium bucket limit! It must be between 1 and 100.",
      unprocessable_entity: "We couldn't process your request. At this time, we can't offer more details, but if the issue persists, please get in touch.",
      missing_required_params: "You are missing a required parameter. Please check your request and, or, consult the API documentation for more details.",
      unhandled_error_code: 'Hello, Dora the explorer! It looks like you managed to find a new error. Please try again and if the issue persists, get in touch.',
      not_implemented: "This endpoint likely doesn't exist or hasn't been implemented.",
      external_error: "The API you requested returned an error. If there is a message available, we'll show it. If the issue persits, get in touch."
    }

    before_action :token, :authority
    skip_before_action :verify_authenticity_token

    def token
      token = Rails.env.development? ? (params['token'] || request.headers['token']) : request.headers['token']
      
      render(
        json: {
          request_id: SecureRandom.uuid,
          code: 'errors.games.directory.common.authentication.missing_token',
          message: ERRORS[:missing_token]
        },
        status: 400
      ) unless token and return token
    end

    def authority
      User.find_by!(uuid: token)
    rescue
      render(
        json: {
          request_id: SecureRandom.uuid,
          code: 'errors.games.directory.common.authentication.invalid_or_expired_token',
          message: ERRORS[:invalid_or_expired_token]
        },
        status: 401
      )
    end

  private

    def respond_with(data = {}, status = 200)
      render(
        status: status,
        json: {
          request_id: SecureRandom.uuid,
          authority: {
            email: authority.email,
            ring: authority.ring,
            limit: 0,
            requests: 0,
            created_at: authority.created_at
          },
          **data
        }
      )
    end

    def reject_with(status, code, server_error = nil)
      Rails.logger.error(server_error) if server_error
      
      respond_with(
        {
          code: "errors.games.directory.#{ code || 'unhandled_error_code' }",
          message: ERRORS[(code.to_s.split('.').last || 'unhandled_error_code').to_sym]
        },
        status
      )      
    end
  end
end
```

Then each action in the API should be wrapped with `respond_with`

```
module API::PlayStation::V1
  class GamesController < API::PlayStation::BaseController
    def index
      limit = (params['per_page'] || 25).to_i

      raise '' if limit.to_i > 100

      concepts = PlayStation::Concept.all
      concepts = concepts.where('name LIKE ?', "%#{ params[:query] }%") if params[:query]

      pagy, games = pagy(
        concepts.order(release_date: :desc),
        items: limit,
        page: params[:page]
      )

      pagy_headers_merge(pagy)

      respond_with(
        games: games.as_json(
          except: [:created_at, :updated_at],
          include: {
            cover: {
              only: [:id, :created_at],
              methods: :url
            },
            background: {
              only: [:id, :created_at],
              methods: :url
            }
          }
        ),
        pagination: pagy_headers_hash(pagy)
      )
    rescue
      reject_with(400, 'common.limits.per_page_exceeded')
    end

  private

    def game
      PlayStation::Concept.friendly.find(params[:id])
    end
  end
end
```